hw/ip/otbn/doc/_index.md

```diff
@@ -0,0 +1,2525 @@
+      </td>
+    </tr>
+    <tr>
+      <td>0x7D4</td>
+      <td>RW</td>
+      <td>
+        <strong>MOD4</strong>.
+        Bits [159:128] of the modulus operand, used in the BN.ADDM/BN.SUBM instructions.
+        This CSR is mapped to the RND WSR.
+      </td>
+    </tr>
+    <tr>
+      <td>0x7D5</td>
+      <td>RW</td>
+      <td>
+        <strong>MOD5</strong>.
+        Bits [191:160] of the modulus operand, used in the BN.ADDM/BN.SUBM instructions.
+        This CSR is mapped to the RND WSR.
+      </td>
+    </tr>
+    <tr>
+      <td>0x7D6</td>
+      <td>RW</td>
+      <td>
+        <strong>MOD6</strong>.
+        Bits [223:192] of the modulus operand, used in the BN.ADDM/BN.SUBM instructions.
+        This CSR is mapped to the RND WSR.
+      </td>
+    </tr>
+    <tr>
+      <td>0x7D7</td>
+      <td>RW</td>
+      <td>
+        <strong>MOD7</strong>.
+        Bits [255:224] of the modulus operand, used in the BN.ADDM/BN.SUBM instructions.
+        This CSR is mapped to the RND WSR.
+      </td>
+    </tr>
+    <tr>
+      <td>0xFC0</td>
+      <td>R</td>
+      <td>
+        <strong>RND</strong>.
+        A random number.
+      </td>
+    </tr>
+  </tbody>
+</table>
+
+### Wide Data Registers
+
+In addition to the 32b wide GPRs, OTBN has a second "wide" register file, which is used by the big number instruction subset.
+This register file consists of NWDR = 32 Wide Data Registers (WDRs).
+Each WDR is WLEN = 256b wide.
+
+| Register |
+|----------|
+| w0       |
+| w1       |
+| ...      |
+| w31      |
+
+### Wide Special Purpose Registers (WSRs)
+
+In addition to the Wide Data Registers, BN instructions can also access WLEN-sized special purpose registers, short WSRs.
+
+<table>
+  <thead>
+    <tr>
+      <th>Number</th>
+      <th>Privilege</th>
+      <th>Description</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <td>0x1</td>
+      <td>RW</td>
+      <td>
+        <strong>MOD</strong>
+        Modulus. To be used in the BN.ADDM and BN.SUBM instructions.
+      </td>
+    </tr>
+    <tr>
+      <td>0x2</td>
+      <td>R</td>
+      <td>
+        <strong>RND</strong>
+        A random number.
+      </td>
+    </tr>
+  </tbody>
+</table>
+
+### Flags
+
+In addition to the wide register file, OTBN maintains global state in two groups of flags for the use by wide integer operations.
+Flag groups are named Flag Group 0 (`FG0`), and Flag Group 1 (`FG1`).
+Each group consists of four flags.
+Each flag is a single bit.
+
+- `C` (Carry flag).
+  Set to 1 an overflow occurred in the last arithmetic instruction.
+
+- `L` (LSb flag).
+  The least significant bit of the result of the last arithmetic or shift instruction.
+
+- `M` (MSb flag)
+  The most significant bit of the result of the last arithmetic or shift instruction.
+
+- `Z` (Zero Flag)
+  Set to 1 if the result of the last operation was zero; otherwise 0.
+
+### Loop Stack
+
+The LOOP instruction allows for nested loops; the active loops are stored on the loop stack.
```

Comment:
I don't want to get into the implementation details in this spec, but should we indicate that CFI/other integrity checks can/will be applied to the loop stack?  (The only implementation detail that might be exposed here would be if we need to have some value-and-its-complement representation of the count, but we might formulate that internally as well.)

Also, since we make recommendations about GPR allocations for code authors in this document, should we make a recommendation that authors use the LOOP construct for data manipulation sequences?  (As opposed to the BEQ/BNE/JAL/JALR instructions?)  I am assuming that LOOP is going to be the most FI-robust structure we have, but perhaps that is not a correct assumption on my part.

Reply:
I know there's a lot more to be said (and discussed first) about how to best use OTBN in a secure fashion, but it certainly makes sense to include our current thinking into the spec already as "preliminary guidance".

@cdgori would you prefer such content to be part of the ISA spec, or would you look for it in a "(Secure) programmer's guide" section further down?

Reply:
I did file https://github.com/lowRISC/opentitan/issues/2499 and https://github.com/lowRISC/opentitan/issues/2498 to keep track of these issues. I'm not sure what we can add to the spec currently, as we want to make sure that we give the right advice, but we also know that there might be changes to the way loops works. If you're ok with it, I'd keep it as a TODO for now (and have added a comment to the doc mentioning that to our readers).

Reply:
Thanks for filing the issues, mostly I was flagging things we need to remember to deal with as the spec/design matures.

Probably some "(secure) programmer's guide" will make sense ultimately, but it might be premature for us to write that yet.  Once we know what countermeasures/features we add it will be clear what the preferred coding styles are.